### PR TITLE
Allow AES XTS decryption using duplicate keys.

### LINKED ISSUE #9105: Unable to decrypt AES-XTS data that uses duplicate key
Seen with OpenSSL 1.1.1c

https://github.com/openssl/openssl/blob/887e22dd8b6f054e39b2d20fc8870eaba7fc61a8/crypto/evp/e_aes.c#L3463

This code disallows setting a duplicate key, however this also implies that openssl can no longer be used to decrypt data that was encrypted with a duplicate key.

Please allow duplicate keys for decryption or advice on how to work around this limitation.

### PR DESCRIPTION
**Pull request by paulidale**
Encryption is still disallowed and will generate an error.

Fixes #9105

### CODE REVIEW

**Comment by paulidale** [comment]
Thanks @mspncp I'll wait to see if @slontis or @davidmakepeace can provide insight as to the FIPS status of this.

**Comment by slontis** [comment]
Hmm.. This looks good to me.
It seems to be a general pattern to have restrictions on protection, but then relax these restrictions during processing.

BUT..

If you look at FIPS 1402IG it specifically says this in the following section..
A.9 XTS-AES Key Generation Requirements

additional comments:

The check for Key_1 ≠ Key_2 shall be done at any place BEFORE using the keys in the XTS-AES algorithm to process data with them.

Maybe we should always do the check during generation/encryption.. but only do the decryption check in FIPS mode??

**Comment by slontis** [comment]
Either way we do need some. mechanism that allows them to decrypt using duplicate keys, otherwise they cant migrate them.. Could also make it configurable, and turn it on in fips and off in default by default (with the ability to change it).  Maybe ue a security flags setter/getter?

**Comment by richsalz** [comment]
IF it all possible, don't make the FIPS and non-FIPS default behavior different.  This probably means requiring a flag for decrypt to override the "no duplicate key" behavior.

**Comment by paulidale** [comment]
Unless @davidmakepeace has a definite opinion about this, I'll ask the experts....

**Comment by davidmakepeace** [comment]
I think we should disable both encryption and decryption using duplicate keys by default for both default and fips providers with a way to re-enable decryption for migration purposes.

Even if there is a popular Application out there in the wild that uses OpenSSL, uses AES XTS and has been deliberately generating duplicate keys I think it is reasonable to prevent all use of duplicate keys by default - such an Application could be updated to perform the migration using a new mechanism to allow decryption for that purpose.

The main use case for this algorithm is encryption of block-oriented storage for databases, filesystems, etc., especially when retrofitted to existing systems where there is no space for metadata such as per-block nonces and authentication tags. I expect that most such systems are read/write though some may be storage archives. In some cases the amount of data may be so large that re-keying/migrating would be onerous and it would be preferable to just keep using the duplicate keys assuming that the risks of not migrating are accepted and well understood. Maybe there should be a mechanism to explicitly allow both encryption and decryption.

It would nice to know if any such storage systems (that have been deliberately generating duplicate keys) exist.

**Comment by paulidale** [comment]
I've updated this PR to disallow duplicate key XTS by default.
I've added a ctrl to allow decryption to be performed.

(documentation is lacking until the final solution is determined)

**Comment by slontis** [comment]
I don't see many alternatives to the way you have written this.

**Comment by paulidale** [comment]
Ping?  A re-review is required after changes.

**Comment by paulidale** [comment]
After hearing from the FIPS lab, decryption is not permitted at all in FIPS mode.

> Unfortunately, as your pointed out the IG states before using keys in the XTS-AES Algorithm to process data with them. Our interpretation in the past has been that this means it applies to both encryption and decryption functions. Other areas in CMVP guidance they explicitly call out encryption vs. decryption, for example the AES GCM IV uniqueness requirements in IG A.5.

The choice is between having a behavioural difference between FIPS and not, or not allowing decryption at all.  I think the behavioural difference is the better path.  Once this decision is made, the requirement for a control to enable decryption doesn't seem so relevant.

File: test/recipes/30-test_evp_data/evpciph.txt [hunk 1]
```diff
@@ -1191,12 +1191,21 @@ Result = CIPHERFINAL_ERROR
...
 
 Cipher = aes-128-xts
+Operation = ENCRYPT
 Key = 0000000000000000000000000000000000000000000000000000000000000000
 IV = 00000000000000000000000000000000
 Plaintext = 0000000000000000000000000000000000000000000000000000000000000000
 Ciphertext = 917cf69ebd68b2ec9b9fe9a3eadda692cd43d2f59598ed858c02c2652fbf922e
 Result = KEY_SET_ERROR
 
+# This is banned under FIPS
```
Comment (mattcaswell) [inline test/recipes/30-test_evp_data/evpciph.txt]: It might be nice to say what specifically is banned (i.e. both keys are the same)

Comment (mattcaswell) [inline test/recipes/30-test_evp_data/evpciph.txt]: It might be nice to say what specifically is banned (i.e. both keys are the same)

Comment (paulidale) [reply test/recipes/30-test_evp_data/evpciph.txt]: Good point.  Once the evp_tests know aboute FIPS, it will be moot but I'll add something in the meantime.

**Comment by richsalz** [comment]
I think a behavior change is *very very wrong.*  Portable applications will mis-behave depending on how the local site is configured.  Make it explicit to always decrypt with duplicate keys.

**Comment by paulidale** [comment]
> Make it explicit to always decrypt with duplicate keys.

Which would mean that we **_cannot_** include this in the FIPS provider.

The choice is between:

1. Have this behavourial difference and have XTS in the FIPS validation.
1. Always allow duplicated key decryption and not have XTS in the FIPS validation.
1. Ban duplicated key decryption and have XTS in the FIPS validation.

My vote is overwhelming for the first of these.
It can't be in FIPS but it is wanted for legacy reasons.

File: test/recipes/30-test_evp_data/evpciph.txt (hunk 1 above)
Comment (paulidale) [reply test/recipes/30-test_evp_data/evpciph.txt]: Added some additional comments.
I'm not intending to add the FIPS mode support at this point, I don't believe that a definitive decision as to how this will be done has yet been made.

**Comment by t-j-h** [comment]
Technical enforcement is not required. We can add this into the security policy passing the responsibility to the user of the module.

**Comment by paulidale** [comment]
I view technical enforcement as _better_ than policy enforcement, especially when it's straightforward like here.

**Comment by slontis** [comment]
> We can add this into the security policy passing the responsibility to the user of the module.
I disagree with this approach. It assumes that EVERY user will READ and actually UNDERSTAND what is in the security policy. I think it is much better for it to fall over and then at least the question gets asked as to why it doesnt work. If you really want to do something insecure it should be an option that you need to enable. The name of the option should make it clear that this is insecure - as well as listing the option in the security policy.

**Comment by richsalz** [comment]
I was thinking of the openssl application having a flag.

I think adding a control that says "allow duplicate keys" that the FIPS XTS code could error-return would work.

**Comment by paulidale** [comment]
That would constitute a FIPS behavioural difference, just a different one.

I don't see a need but I can add a derypt outside of FIPS mode control back in if it is desirable, I think it is more complex than required and requires code changes in applications to support legacy use.

**Comment by richsalz** [comment]
Here is what I am suggesting, hopefully this is more clear.  Add a CTRL option that says "can decrypt with duplicate keys"  FIPS provider rejects it, default provider allows it.  Applications that want to do this will/can/should know that the default provider will decrypt. The decrypt applet must be also be changed: add a new flag "-dupkeys" or whatever, and if that is specified, it explicitly loads the default provider to do the decryption regardless of whatever providers are loaded by switches or config files.

Is that more helpful?

**Comment by paulidale** [comment]
XTS mode isn't supported by enc, so no need to modify the applet.

By introducing a new ctl call, _user_ applications will have to change to support it.  By doing the way it is done here, they don't.  BTW: the first commit in this branch does include a new ctl call to enable same key decryption.  I think it is close to what you're suggesting.  The flawed assumption being that this was allow under the FIPS legacy rules.

**Comment by mattcaswell** [comment]
> Have this behavourial difference and have XTS in the FIPS validation.

Seems like the right answer to me. Behavioural differences between the default and FIPS providers are inevitable so I have no problem with that at all.

**Comment by paulidale** [comment]
Ping for review?

File: crypto/evp/e_aes.c [hunk 2]
```diff
@@ -401,7 +408,8 @@ static int aesni_xts_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
          * This addresses Rogaway's vulnerability.
          * See comment in aes_xts_init_key() below.
          */
-        if (memcmp(key, key + bytes, bytes) == 0) {
+        if ((!allow_insecure_decrypt || enc)
+                && memcmp(key, key + bytes, bytes) == 0) {
```
Comment (davidben) [inline crypto/evp/e_aes.c]: Existing issue and just a single drive-by comment, but surely this should be a constant-time `memcmp`. :-)

(Ditto throughout.)

Comment (paulidale) [reply crypto/evp/e_aes.c]: Maybe??

I'm having trouble thinking what's going to leak.  The number of common bytes of the two keys?

I'll update just to be sure.

**Comment by paulidale** [comment]
Travis failure isn't relevant.

**Comment by paulidale** [comment]
Ping for review?

**Comment by paulidale** [comment]
Thanks @t8m

Merged to master and 1.1.1.
